cardano-crypto-praos/src/Cardano/Crypto/VRF/Praos.hs

```diff
@@ -0,0 +1,464 @@
+  --
+  -- Sizes of various value types involved in the VRF calculations. Users of
+  -- this module will not need these, we are only exporting them for unit
+  -- testing purposes.
+  , crypto_vrf_proofbytes
+  , crypto_vrf_publickeybytes
+  , crypto_vrf_secretkeybytes
+  , crypto_vrf_seedbytes
+  , crypto_vrf_outputbytes
+
+  -- * Value types
+  --
+  -- These types are all implemented as transparent references. The actual
+  -- values are kept entirely in C memory, allocated when a value is created,
+  -- and freed when the value's finalizer runs.
+  , Seed
+  , SK
+  , PK
+  , Proof
+  , Output
+
+
+  -- * Key sizes
+  , certSizeVRF
+  , signKeySizeVRF
+  , verKeySizeVRF
+  , vrfKeySizeVRF
+
+  -- * Seed and key generation
+  , genSeed
+  , keypairFromSeed
+
+  -- * Conversions
+  , unsafeRawSeed
+  , outputBytes
+  , proofBytes
+  , skBytes
+  , pkBytes
+  , skToPK
+  , skToSeed
+
+  -- * Core VRF operations
+  , prove
+  , verify
+  
+  , SignKeyVRF (..)
+  , VerKeyVRF (..)
+  , CertVRF (..)
+  )
+where
+
+import Cardano.Binary
+  ( FromCBOR (..)
+  , ToCBOR (..)
+  , serialize'
+  )
+
+import Cardano.Crypto.VRF.Class
+import Cardano.Prelude (NoUnexpectedThunks, UseIsNormalForm(..))
+import Cardano.Crypto.Seed (getBytesFromSeedT)
+import GHC.Generics (Generic)
+import Data.Coerce (coerce)
+
+import Foreign.ForeignPtr
+import Foreign.C.Types
+import Foreign.Ptr
+import Foreign.Marshal.Alloc
+import Foreign.Marshal.Utils
+import System.IO.Unsafe (unsafePerformIO)
+import Control.Monad (void)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import Data.Maybe (isJust)
+import Data.Proxy (Proxy (..))
+
+-- Value types.
+--
+-- These are all transparent to the Haskell side of things, all we ever do
+-- with these is pass pointers to them around. We don't want to know anything
+-- about them, hence, we make them uninhabited (isomorphic with
+-- 'Data.Void.Void'). The reason we have them at all, rather than duplicating
+-- C's void pointers, is because we want to distinguish them at the type
+-- level.
+
+data SeedValue
+data SKValue
+data PKValue
+data ProofValue
+data OutputValue
+
+-- Type aliases for raw pointers
+--
+-- These will not leave this module, they are only here for our convenience,
+-- so we can afford to not newtype them.
+
+type SeedPtr = Ptr SeedValue
+type SKPtr = Ptr SKValue
+type PKPtr = Ptr PKValue
+type ProofPtr = Ptr ProofValue
+type OutputPtr = Ptr OutputValue
+
+-- The exported types.
+--
+-- These are wrappers around 'ForeignPtr's; we don't export the constructors,
+-- so callers have to go through our blessed API to create any of them. This
+-- way we can make sure that we always allocate the correct sizes, and attach
+-- finalizers that automatically free the memory for us.
+
+-- | A random seed, used to derive a public/secret key pair.
+newtype Seed = Seed { unSeed :: ForeignPtr SeedValue }
+
+-- | Secret key. In this implementation, the secret key is actually a 64-byte
```

Comment:
If these are to be exported, we have generally gone with `Sign`/`Verify` rather than `Private`/`Public`. So better to be consistent here.

Reply:
In that case, let's also change identifiers to reflect this naming conversion, at least up to the FFI boundary. So e.g. `PKValue` -> `VerKeyValue`, etc.

Reply:
I think it's not too bad if it's internal, since `libsodium` does use the SK/PK convention. But I'm also happy if you want to change them

Reply:
I changed them up to the FFI boundary, that is, the functions imported directly from libsodium follow libsodium naming conventions, while the "nicer" wrappers use our conventions. Seems like the most logical approach to me.